File: Redcoder/Python/62.py
H, W = map(int,input().split())
box = []
for i in range(10):
    b = list(map(int,input().split()))
    box.append(b)

for k in range(10):
    for i in range(10):
        for j in range(10):
            box[i][j] = min(box[i][j], box[i][k] + box[k][j])

box2 = []
for i in range(H):
    a = list(map(int,input().split()))
    box2.append(a)
ans = 0
for i in range(H):
    for j in range(W):
        if box2[i][j] == -1:
            continue
        ans += box[box2[i][j]][1]
print(ans)

File: Redcoder/C++/62.cpp
#include <bits/stdc++.h>
#define rep(i,nn,n) for (int i = nn; i < (n); ++i)
#define rrep(i,nn,n) for (int i = nn; i > (n); --i)
using namespace std;
using ll = long long;
using P = pair<int,int>;
#define chmin(x,y) x = min(x,y)
#define chmax(x,y) x = max(x,y)



int main() {
    int H,W;
    cin >> H >> W;
    vector<vector<int>> box(10,vector<int>(10,0));
    rep(i,0,10){
        rep(j,0,10){
            cin >> box[i][j];
        } 
    }
    

    rep(k,0,10){
        rep(i,0,10){
            rep(j,0,10){
                chmin(box[i][j], box[i][k] + box[k][j]);
            }
        } 
    }

    int box2[10]{0,0,0,0,0,0,0,0,0,0};
    rep(i,0,H) {
        rep(j,0,W){
            int a;
            cin >> a;
            ++box2[a];
        }
    }
    int ans = 0;
    rep(i,0,10){
        ans += box[i][1] * box2[i];
    }


    cout << ans << endl;
    return 0;
}